# Домашнее задание

В домашнем задании вам необходимо рассмтреть данные с олимпиады НТИ по профилю "Исскусственный интеллект" за 2019-2020 год.

Задачу можно найти по ссылке:

https://www.kaggle.com/c/onti-students-performance/data

## **Как загрузить данные в Google Colab?**

1. Из Github (файлы <25 МБ)
2. С локального диска
3. С Google Диска

In [111]:
# Github
# 1. Нажмите на файл с данными, которые хотите подключить
# 2. Нажмите "Raw"
# 3. Скопируйте ссылку открывшейся старницы
import pandas as pd

url1 = 'https://raw.githubusercontent.com/Adelaaas/Data_science_basic_22-23_1/main/class_work_4_Pandas/ONTI%20Students%20performance/X_train.csv'
url2 = 'https://raw.githubusercontent.com/Adelaaas/Data_science_basic_22-23_1/main/class_work_4_Pandas/ONTI%20Students%20performance/studs_info.csv'
url3 = 'https://raw.githubusercontent.com/Adelaaas/Data_science_basic_22-23_1/main/class_work_4_Pandas/ONTI%20Students%20performance/zachety.csv'
url4 = 'https://raw.githubusercontent.com/Adelaaas/Data_science_basic_22-23_1/main/class_work_4_Pandas/ONTI%20Students%20performance/y_train.csv'

X_train = pd.read_csv(url1)
studs_info = pd.read_csv(url2)
zachety = pd.read_csv(url3)
y_train = pd.read_csv(url4)

In [112]:
studs_info

STD_ID     number Пол Статус Дата выпуска Категория обучения  \
0      99136    13951.0   Ж    ВЫП   2019-08-31                 БП   
1      99112    13998.0   Ж    ВЫП   2019-08-31                 БП   
2      99109    13999.0   Ж    ВЫП   2019-08-31                 БП   
3      99125    13953.0   Ж    ВЫП   2019-08-31                 БП   
4     100676    13950.0   М    ВЫП   2019-08-31                 БП   
...      ...        ...  ..    ...          ...                ...   
5461  100678      978.0   Ж     СТ          NaN                 БП   
5462  100663     1021.0   Ж     СТ          NaN                ДОГ   
5463  191630     6447.0   Ж     СТ          NaN                ДОГ   
5464  191625     3765.0   Ж     СТ          NaN                ДОГ   
5465  194978     5245.0   Ж     СТ          NaN                ДОГ   

     Форма обучения      Шифр направление (специальность)         Образование  \
0                 Д  31.05.01               Лечебное дело  2629.0          СР   
1                 Д  31.05.01               Лечебное дело  8240.0          СР   
2                 Д  31.05.01               Лечебное дело  2314.0          СР   
3                 Д  31.05.01               Лечебное дело  2345.0          СР   
4                 Д  31.05.01               Лечебное дело  2650.0          СР   
...             ...       ...                         ...     ...         ...   
5461              Д  31.05.01               Лечебное дело  2388.0          СР   
5462              Д  31.05.01               Лечебное дело  2554.0          СР   
5463              Д  31.05.01               Лечебное дело  6692.0          СР   
5464              Д  31.05.01               Лечебное дело     NaN         NaN   
5465              Д  31.05.01               Лечебное дело     NaN         NaN   

              Дата выдачи        Что именно закончил  
0     2013-06-26 00:00:00  ИНОГОРОДНЯЯ СРЕДНЯЯ ШКОЛА  
1     2012-06-23 00:00:00     МОУ СОШ N   2, ПЕСТОВО  
2     2013-06-25 00:00:00  ИНОГОРОДНЯЯ СРЕДНЯЯ ШКОЛА  
3     2013-06-23 00:00:00        Гимназия "ГАРМОНИЯ"  
4     2012-06-29 00:00:00  ИНОГОРОДНЯЯ СРЕДНЯЯ ШКОЛА  
...                   ...                        ...  
5461  2013-06-21 00:00:00  ИНОГОРОДНЯЯ СРЕДНЯЯ ШКОЛА  
5462  2013-06-27 00:00:00  ИНОГОРОДНЯЯ СРЕДНЯЯ ШКОЛА  
5463  2014-05-31 00:00:00  ИНОГОРОДНЯЯ СРЕДНЯЯ ШКОЛА  
5464                  NaN                        NaN  
5465                  NaN                        NaN  

[5466 rows x 13 columns]

In [18]:
1 
print(X_train.shape)
print(studs_info.shape)
print(zachety.shape)
print(y_train.shape)

(105058, 8)
(5466, 13)
(133223, 8)
(105058, 2)


In [124]:
2 
train = pd.concat([y_train, X_train],axis=1)

In [11]:
train

index  mark  Unnamed: 0  STD_ID             НАПРАВЛЕНИЕ        ГОД  \
0        51081     5       51081  175711     38.03.02 Менеджмент  2018-2019   
1        72921     5       72921  100647  31.05.01 Лечебное дело  2015-2016   
2        96438     5       96438  199311   31.05.03 Стоматология  2019-2020   
3       273044     4      273044  162002     38.03.02 Менеджмент  2017-2018   
4       371309     3      371309  108384  31.05.01 Лечебное дело  2018-2019   
...        ...   ...         ...     ...                     ...        ...   
105053  224856     3      224856  176141  40.03.01 Юриспруденция  2017-2018   
105054  260998     3      260998  182794  31.05.01 Лечебное дело  2018-2019   
105055   57245     4       57245  162391  31.05.01 Лечебное дело  2016-2017   
105056  218529     4      218529  202479      38.03.01 Экономика  2017-2018   
105057  233318     4      233318  182325  08.03.01 Строительство  2018-2019   

        КУРС  СЕМЕСТР                АТТЕСТАЦИЯ  \
0        2.0      4.0                   Экзамен   
1        3.0      5.0                   Экзамен   
2        1.0      1.0                   Экзамен   
3        2.0      3.0                   Экзамен   
4        5.0      9.0                   Экзамен   
...      ...      ...                       ...   
105053   1.0      2.0           Курсовая работа   
105054   2.0      4.0                   Экзамен   
105055   1.0      2.0  Дифференцированный зачет   
105056   1.0      1.0  Дифференцированный зачет   
105057   2.0      3.0           Курсовая работа   

                                               ДИСЦИПЛИНА  
0                                               Маркетинг  
1       Пропедевтика внутренних болезней, лучевая диаг...  
2                                                   Химия  
3                                              Статистика  
4                                               Педиатрия  
...                                                   ...  
105053                              Конституционное право  
105054                              Нормальная физиология  
105055                                   Иностранный язык  
105056                                        Информатика  
105057           Технологические процессы в строительстве  

[105058 rows x 10 columns]

In [122]:
3 
train.isna().sum()


index          0
mark           0
Unnamed: 0     0
STD_ID         0
НАПРАВЛЕНИЕ    0
ГОД            0
КУРС           0
СЕМЕСТР        0
АТТЕСТАЦИЯ     0
ДИСЦИПЛИНА     0
dtype: int64

In [123]:
 4
 print(studs_info['направление (специальность)'].unique())

print(studs_info.value_counts('направление (специальность)'))

['Лечебное дело' 'Фундаментальная и прикладная химия' 'Стоматология'
 'Фармация' 'Экономика' 'Менеджмент'
 'Информатика и вычислительная техника'
 'Конструкторско-технологическое обеспечение машиностроительных производств'
 'Лингвистика' 'Управление персоналом' 'Агрономия' 'Биология'
 'Лесное дело' 'Радиотехника' 'Зоотехния' 'Строительство'
 'Экология и природопользование' 'Агроинженерия' 'Журналистика'
 'Землеустройство и кадастры' 'Государственное и муниципальное управление'
 'Юриспруденция' 'Теплоэнергетика и теплотехника'
 'Электроника и наноэлектроника' 'Филология'
 'Конструирование и технология электронных средств'
 'Прикладная математика и информатика' 'Дизайн'
 'Технология художественной обработки материалов'
 'Мехатроника и робототехника'
 'Педагогическое образование (с двумя профилями подготовки)' 'Архитектура'
 'Дизайн архитектурной среды' 'Педагогическое образование'
 'Психолого-педагогическое образование' 'История'
 'Эксплуатация транспортно-технологических машин и комплек

In [50]:
5
new_df = studs_info. copy ()
a=new_df[new_df["Пол"] == "Ж"] 
b=a[a["направление (специальность)"] == "Лечебное дело"]
b

STD_ID     number Пол Статус Дата выпуска Категория обучения  \
0      99136    13951.0   Ж    ВЫП   2019-08-31                 БП   
1      99112    13998.0   Ж    ВЫП   2019-08-31                 БП   
2      99109    13999.0   Ж    ВЫП   2019-08-31                 БП   
3      99125    13953.0   Ж    ВЫП   2019-08-31                 БП   
5      99099    13959.0   Ж    ВЫП   2019-08-31                 БП   
...      ...        ...  ..    ...          ...                ...   
5461  100678      978.0   Ж     СТ          NaN                 БП   
5462  100663     1021.0   Ж     СТ          NaN                ДОГ   
5463  191630     6447.0   Ж     СТ          NaN                ДОГ   
5464  191625     3765.0   Ж     СТ          NaN                ДОГ   
5465  194978     5245.0   Ж     СТ          NaN                ДОГ   

     Форма обучения      Шифр направление (специальность)         Образование  \
0                 Д  31.05.01               Лечебное дело  2629.0          СР   
1                 Д  31.05.01               Лечебное дело  8240.0          СР   
2                 Д  31.05.01               Лечебное дело  2314.0          СР   
3                 Д  31.05.01               Лечебное дело  2345.0          СР   
5                 Д  31.05.01               Лечебное дело  8800.0          СР   
...             ...       ...                         ...     ...         ...   
5461              Д  31.05.01               Лечебное дело  2388.0          СР   
5462              Д  31.05.01               Лечебное дело  2554.0          СР   
5463              Д  31.05.01               Лечебное дело  6692.0          СР   
5464              Д  31.05.01               Лечебное дело     NaN         NaN   
5465              Д  31.05.01               Лечебное дело     NaN         NaN   

              Дата выдачи        Что именно закончил  
0     2013-06-26 00:00:00  ИНОГОРОДНЯЯ СРЕДНЯЯ ШКОЛА  
1     2012-06-23 00:00:00     МОУ СОШ N   2, ПЕСТОВО  
2     2013-06-25 00:00:00  ИНОГОРОДНЯЯ СРЕДНЯЯ ШКОЛА  
3     2013-06-23 00:00:00        Гимназия "ГАРМОНИЯ"  
5     2013-06-21 00:00:00     МОУ СОШ (полная), ХОЛМ  
...                   ...                        ...  
5461  2013-06-21 00:00:00  ИНОГОРОДНЯЯ СРЕДНЯЯ ШКОЛА  
5462  2013-06-27 00:00:00  ИНОГОРОДНЯЯ СРЕДНЯЯ ШКОЛА  
5463  2014-05-31 00:00:00  ИНОГОРОДНЯЯ СРЕДНЯЯ ШКОЛА  
5464                  NaN                        NaN  
5465                  NaN                        NaN  

[564 rows x 13 columns]

In [115]:
6
c= studs_info['Дата выпуска'].str.split('-',expand=True)

c.columns=['Год выпуска','Месяц выпуска','День выпуска']
trainc= pd.concat([studs_info, c],axis=1)



STD_ID     number Пол Статус Категория обучения Форма обучения  \
0      99136    13951.0   Ж    ВЫП                  0              Д   
1      99112    13998.0   Ж    ВЫП                  0              Д   
2      99109    13999.0   Ж    ВЫП                  0              Д   
3      99125    13953.0   Ж    ВЫП                  0              Д   
4     100676    13950.0   М    ВЫП                  0              Д   
...      ...        ...  ..    ...                ...            ...   
5461  100678      978.0   Ж     СТ                  0              Д   
5462  100663     1021.0   Ж     СТ                ДОГ              Д   
5463  191630     6447.0   Ж     СТ                ДОГ              Д   
5464  191625     3765.0   Ж     СТ                ДОГ              Д   
5465  194978     5245.0   Ж     СТ                ДОГ              Д   

          Шифр направление (специальность)         Образование  \
0     31.05.01               Лечебное дело  2629.0          СР   
1     31.05.01               Лечебное дело  8240.0          СР   
2     31.05.01               Лечебное дело  2314.0          СР   
3     31.05.01               Лечебное дело  2345.0          СР   
4     31.05.01               Лечебное дело  2650.0          СР   
...        ...                         ...     ...         ...   
5461  31.05.01               Лечебное дело  2388.0          СР   
5462  31.05.01               Лечебное дело  2554.0          СР   
5463  31.05.01               Лечебное дело  6692.0          СР   
5464  31.05.01               Лечебное дело     NaN         NaN   
5465  31.05.01               Лечебное дело     NaN         NaN   

              Дата выдачи        Что именно закончил Год выпуска  \
0     2013-06-26 00:00:00  ИНОГОРОДНЯЯ СРЕДНЯЯ ШКОЛА        2019   
1     2012-06-23 00:00:00     МОУ СОШ N   2, ПЕСТОВО        2019   
2     2013-06-25 00:00:00  ИНОГОРОДНЯЯ СРЕДНЯЯ ШКОЛА        2019   
3     2013-06-23 00:00:00        Гимназия "ГАРМОНИЯ"        2019   
4     2012-06-29 00:00:00  ИНОГОРОДНЯЯ СРЕДНЯЯ ШКОЛА        2019   
...                   ...                        ...         ...   
5461  2013-06-21 00:00:00  ИНОГОРОДНЯЯ СРЕДНЯЯ ШКОЛА         NaN   
5462  2013-06-27 00:00:00  ИНОГОРОДНЯЯ СРЕДНЯЯ ШКОЛА         NaN   
5463  2014-05-31 00:00:00  ИНОГОРОДНЯЯ СРЕДНЯЯ ШКОЛА         NaN   
5464                  NaN                        NaN         NaN   
5465                  NaN                        NaN         NaN   

     Месяц выпуска День выпуска  
0               08           31  
1               08           31  
2               08           31  
3               08           31  
4               08           31  
...            ...          ...  
5461           NaN          NaN  
5462           NaN          NaN  
5463           NaN          NaN  
5464           NaN          NaN  
5465           NaN          NaN  

[5466 rows x 15 columns]

In [71]:
7 
print(studs_info['Категория обучения'].unique())

['БП' 'ДОГ' nan]


In [107]:
studs_info

0

In [116]:
8
studs_info.loc[(studs_info['Категория обучения'] == 'БП'), 'Категория обучения'] = 0
studs_info.loc[(studs_info['Категория обучения'] == 'ДОГ'),'Категория обучения'] = 1


In [118]:
9 
studs_info.drop(labels=["Дата выдачи"],axis=1)


STD_ID     number Пол Статус Дата выпуска Категория обучения  \
0      99136    13951.0   Ж    ВЫП   2019-08-31                  0   
1      99112    13998.0   Ж    ВЫП   2019-08-31                  0   
2      99109    13999.0   Ж    ВЫП   2019-08-31                  0   
3      99125    13953.0   Ж    ВЫП   2019-08-31                  0   
4     100676    13950.0   М    ВЫП   2019-08-31                  0   
...      ...        ...  ..    ...          ...                ...   
5461  100678      978.0   Ж     СТ          NaN                  0   
5462  100663     1021.0   Ж     СТ          NaN                  1   
5463  191630     6447.0   Ж     СТ          NaN                  1   
5464  191625     3765.0   Ж     СТ          NaN                  1   
5465  194978     5245.0   Ж     СТ          NaN                  1   

     Форма обучения      Шифр направление (специальность)         Образование  \
0                 Д  31.05.01               Лечебное дело  2629.0          СР   
1                 Д  31.05.01               Лечебное дело  8240.0          СР   
2                 Д  31.05.01               Лечебное дело  2314.0          СР   
3                 Д  31.05.01               Лечебное дело  2345.0          СР   
4                 Д  31.05.01               Лечебное дело  2650.0          СР   
...             ...       ...                         ...     ...         ...   
5461              Д  31.05.01               Лечебное дело  2388.0          СР   
5462              Д  31.05.01               Лечебное дело  2554.0          СР   
5463              Д  31.05.01               Лечебное дело  6692.0          СР   
5464              Д  31.05.01               Лечебное дело     NaN         NaN   
5465              Д  31.05.01               Лечебное дело     NaN         NaN   

            Что именно закончил  
0     ИНОГОРОДНЯЯ СРЕДНЯЯ ШКОЛА  
1        МОУ СОШ N   2, ПЕСТОВО  
2     ИНОГОРОДНЯЯ СРЕДНЯЯ ШКОЛА  
3           Гимназия "ГАРМОНИЯ"  
4     ИНОГОРОДНЯЯ СРЕДНЯЯ ШКОЛА  
...                         ...  
5461  ИНОГОРОДНЯЯ СРЕДНЯЯ ШКОЛА  
5462  ИНОГОРОДНЯЯ СРЕДНЯЯ ШКОЛА  
5463  ИНОГОРОДНЯЯ СРЕДНЯЯ ШКОЛА  
5464                        NaN  
5465                        NaN  

[5466 rows x 12 columns]

In [ ]:
# С локального компьютера

from google.colab import files
uploaded = files.upload()

In [ ]:
studs_info = pd.read_csv('studs_info.csv')
X_train = pd.read_csv('X_train.csv')
y_train = pd.read_csv('y_train.csv')
zachety = pd.read_csv('zachety.csv')

In [ ]:
# Google disk

from google.colab import drive
drive.mount('/content/drive')

In [ ]:
path = "copied path"
# /content/drive/MyDrive/Data Science для школьников/ОНТИ /ONTI Students performance/X_train.csv
df_bonus = pd.read_csv(path)

## Вариант 1.

Кто делает:
1. Александров Данила Александрович
2. Быков Дмитрий Александрович 
3. Горбунов Сергей Михайлович 
4. Горячко Дарья Романовна 
5. Громов Алексей Игоревич
6. Игнатов Константин Дмитриевич
7. Ямпольская Ирина Вадимовна
8. Дидковский Илья Радионович

1. Посмотреть размерность дата фреймов.
2. Соедить два датафрейма X_train и y_train в один train по индексам.
3. Посчитать пропуски в получившемся датафрейме train.
4. Посчитать сколько раз встречается каждая специальность в колонке studs_info['направление (специальность)'].
5. Создать новый датафрейм из studs_info, в котором будут только женщины с лечебного дела.
6. Разложить колонку "Дата выпуска" на три: Год выпуска, Месяц выпуска, День выпуска.
7. Посчитать сколько уникальных значений в колонке studs_info['Категория обучения'].
8. Заменить в колонке studs_info['Категория обучения'] все строковые категории на числовые.
9. Удалить ненужную на ваш вгляд колонку из датафрейма studs_info.
10. Посчитать сколько мужчин выпустились 2019-08-31.

## Вариант 2.

Кто делает:
1. Ивлиев Степан Дмитриевич
2. Игнатов Игорь Дмитриевич
3. Казанов Амир Ринатович
4. Клячин Иван Николаевич
5. Котов Александр Андреевич
6. Мелешкин Кирилл Андреевич
7. Файзуллин Илья Денисович
8. Чупрова Ася Александровна

1. Посмотреть размерность дата фреймов.
2. Соедить два датафрейма X_train и y_train в один train по индексам.
3. Посчитать пропуски в датафрейме studs_info.
4. Посчитать сколько мужчин, сколько женщин в датафрейме studs_info.
5. Создать новый датафрейм из studs_info, в котором будут только мужчины 2019-06-30 года выпуска.
6. Разложить колонку "Дата выдачи" на четыре: Год выдачи, Месяц выдачи, День выдачи, Время выдачи.
7. Посчитать сколько уникальных значений в колонке studs_info['Статус'] и сколько раз каждое значение встречается в датафрейм.
8. Посчитать сколько раз какая оценка встречатся в датафрейм y_train.
9. Посчитать средний балл для студента с ID=175711 по всем предметам.
10. Удалить любую ненужны колонку на ваш вгляд из датафрейм train.

In [ ]:
# Ваше решение

df[df['name'] == 'Ivan']

## Вариант 3.

Кто делает:
1. Морозов Максим Олегович
2. Переттруов Никита Артемович
3. Пославский Егор Константинович
4. Садохин Константин Павлович
5. Скиданов Дмитрий Артёмович
6. Судариков Андрей Геннадьевич
7. Таланов Максим Андреевич

Задания:
1. Посмотреть размерность дата фреймов.
2. Соедить два датафрейма X_train и y_train в один train по индексам.
3. Посчитать пропуски в датафрейме studs_info.
4. Посчитать сколько уникальных значений в колонке АТТЕСТАЦИЯ и сколько раз каждое из них встречается.
5. Переименовать все колонки датафрейм train, чтобы названия были маленькими буквами.
6. Разложить колонку "направление" на две: Код направления, Название направления.
7. Создать новый датафрейм, в котором будут только мужчины 3 курса обучения.
8. Заменить все строковые значения в колонке Пол на числовые.
9. Сколько всего студентов сдавали предмет "Информатика" в виде экзамена? И какой средний балл по всем сдававшим?

In [ ]:
# Ваше решение